lib/vector-core/src/fanout.rs

```diff
@@ -63,165 +62,330 @@ impl Fanout {
-                match poll(sink.as_mut(), cx) {
-                    Poll::Pending => poll_result = Poll::Pending,
-                    Poll::Ready(Ok(())) => (),
-                    Poll::Ready(Err(())) => {
-                        self.handle_sink_error(i)?;
-                        continue;
+    /// Send a batch of events to all connected sinks.
+    ///
+    /// This will block on the resolution of any pending reload before proceeding with the send
+    /// operation.
+    ///
+    /// # Panics
+    ///
+    /// This method can panic if the fanout receives a control message that violates some invariant
+    /// about its current state (e.g. remove a non-existant sink, etc). This would imply a bug in
+    /// Vector's config reloading logic.
+    pub async fn send_all(&mut self, events: Vec<Event>) {
+        self.process_control_messages();
+        self.wait_for_replacements().await;
+
+        // The call to `wait_for_replacements` above ensures that all replacement operations are
+        // complete at this point, and we don't need to worry about any of the sinks being `None`.
+        let sink_count = self.sinks.iter().filter(|x| x.1.is_some()).count();
+        debug_assert_eq!(sink_count, self.sinks.len());
+
+        if self.sinks.is_empty() {
+            return;
+        }
+
+        let mut clone_army: Vec<Vec<Event>> = Vec::with_capacity(sink_count);
+        for _ in 0..(sink_count - 1) {
+            clone_army.push(events.clone());
+        }
+        clone_army.push(events);
+
+        let sinks = self
+            .sinks
+            .drain(..)
+            .map(|(id, sink)| (id, sink.expect("no replacements in progress")))
+            .zip(clone_army);
+
+        let mut send_group = SendGroup::with_capacity(sink_count);
+
+        for ((id, sink), events) in sinks {
+            send_group.push(id, sink, events);
+        }
+
+        loop {
+            tokio::select! {
+                sinks = &mut send_group => {
+                    // All in-flight sends have completed, so return sinks to the base collection.
+                    // We extend instead of assign here because other sinks could have been added
+                    // while the send was in-flight.
+                    self.sinks.extend(sinks);
+                    break;
+                }
+                maybe_msg = self.control_channel.recv() => {
+                    match maybe_msg {
+                        Some(msg @ ControlMessage::Add(..)) => {
+                            // New sinks will be picked up on the next call, so we just need to add
+                            // them to the base `sinks` collection.
+                            self.handle_control_message(msg);
+                        },
+                        Some(ControlMessage::Remove(key)) => {
+                            if self.contains(&key) {
+                                // Sink was added while sends were in-flight, so just remove it.
```

Comment:
Should this and below read "were _not_ in-flight"?

Reply:
No, but I can see the confusion. I've reworked all this with the most recent commit and added more thorough comments, so let me know if it's still unclear.